### PR TITLE
eth: add GasPriceOracle

### PR DESCRIPTION
**Pull request by zsfelfoldi**

### CODE REVIEW

File: core/chain_manager.go [hunk 1]
```diff
@@ -419,7 +422,25 @@ func (self *ChainManager) GetBlock(hash common.Hash) *types.Block {
 		glog.V(logger.Error).Infof("invalid block RLP for hash %x: %v", hash, err)
 		return nil
 	}
-	return (*types.Block)(&block)
+
+	blk := (*types.Block)(&block)
```
Comment (obscuren) [inline core/chain_manager.go]: This needs to be moved to its own separate function. We want to limit the amount of database gets and RLP decoding required to get a block. Most cases when `GetBlock` is called you don't require the receipts.

Move this to an additional method like

```go
func (cm *ChainManager) GetReceipts(block *types.Block) types.Receipts {
    // your code
}
```

File: eth/gasprice.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+import (
+	"math/big"
+	"math/rand"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	//"github.com/ethereum/go-ethereum/logger"
```
Comment (obscuren) [inline eth/gasprice.go]: remove

File: core/chain_manager.go [hunk 3]
```diff
@@ -419,7 +422,25 @@ func (self *ChainManager) GetBlock(hash common.Hash) *types.Block {
...
+	rdata, err := self.extraDb.Get(append(receiptsPre, hash[:]...))
+
+	if err == nil {
+		var receipts types.Receipts
+		err = rlp.DecodeBytes(rdata, &receipts)
+		if err != nil {
+			fmt.Println(err)
+		}
+		blk.SetReceipts(receipts)
+		//fmt.Printf("Receipts found for block #%v\n", blk.NumberU64())
```
Comment (obscuren) [inline core/chain_manager.go]: remove comments

File: eth/gasprice.go [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+type GasPriceOracle struct {
+	eth                           *Ethereum
+	chain                         *core.ChainManager
+	pool                          *core.TxPool
+	events                        event.Subscription
+	blocks                        map[uint64]*blockPriceInfo
+	firstProcessed, lastProcessed uint64
+	lastBase                      *big.Int
```
Comment (obscuren) [inline eth/gasprice.go]: Style: group `lastBase` and move the `lbMutex` above. Please rename `lbMutex` to `lastBaseMu` (style throughout the rest of the system)

File: eth/gasprice.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	if bpi == nil {
+		bpi = &blockPriceInfo{}
+		self.blocks[i] = bpi
+	}
+	bpi.baseGasPrice = newBase
+	self.lbMutex.Lock()
+	self.lastBase = newBase
+	self.lbMutex.Unlock()
+
+	//glog.V(logger.Info).Infof("Processed block #%v, base price is %v\n", block.NumberU64(), newBase.Int64())
```
Comment (obscuren) [inline eth/gasprice.go]: uncomment and love to `logger.Detail`

**Comment by tgerring** [comment]
@zsfelfoldi If it's not too much effort, can you add a method to XEth to access the receipts? This would be to support #1042

**Comment by obscuren** [comment]
@zsfelfoldi @tgerring as per my question on gitter. Separate the transaction receipt storing in a new PR

**Comment by zsfelfoldi** [comment]
@tgerring Ok, I'll do that later today or tomorrow morning.

**Comment by zsfelfoldi** [comment]
https://github.com/ethereum/go-ethereum/pull/1049

**Comment by obscuren** [comment]
Mining just empty blocks:

```
11062348998802'
> I0611 00:23:29.796654   84753 chain_manager.go:686] imported 4 block(s) (0 queued 0 ignored) in 8.59473ms. #80 [a6a89a15 / cafa7387]
I0611 00:23:29.797283   84753 worker.go:413] commit new work on block 81 with 0 txs & 1 uncles
> eth.gasPrice
'10957699177273'
> I0611 00:23:30.982363   84753 chain_manager.go:686] imported 1 block(s) (0 queued 0 ignored) in 10.982954ms. #81 [eaf58cd0 / eaf58cd0]
I0611 00:23:30.982507   84753 worker.go:256] 🔨  Mined block (#81 / eaf58cd0). Wait 5 blocks for confirmation
I0611 00:23:30.983298   84753 worker.go:413] commit new work on block 82 with 0 txs & 2 uncles
I0611 00:23:32.064667   84753 chain_manager.go:686] imported 1 block(s) (0 queued 0 ignored) in 14.537861ms. #82 [510b1350 / 510b1350]
I0611 00:23:32.065081   84753 worker.go:256] 🔨  Mined block (#82 / 510b1350). Wait 5 blocks for confirmation
I0611 00:23:32.065972   84753 worker.go:413] commit new work on block 83 with 0 txs & 0 uncles
> eth.gasPriceI0611 00:23:32.350849   84753 chain_manager.go:686] imported 1 block(s) (0 queued 0 ignored) in 5.931852ms. #83 [62056c3b / 62056c3b]
I0611 00:23:32.351061   84753 worker.go:256] 🔨  Mined block (#83 / 62056c3b). Wait 5 blocks for confirmation
I0611 00:23:32.351897   84753 worker.go:413] commit new work on block 84 with 0 txs & 0 uncles

'11845227651911'
> I0611 00:23:33.381041   84753 chain_manager.go:686] imported 1 block(s) (0 queued 0 ignored) in 5.865722ms. #84 [c78e753f / c78e753f]
I0611 00:23:33.381149   84753 worker.go:256] 🔨  Mined block (#84 / c78e753f). Wait 5 blocks for confirmation
I0611 00:23:33.382113   84753 worker.go:413] commit new work on block 85 with 0 txs & 0 uncles
> eth.gasPrice
'11734000964259'
```

Shouldn't it dip below 1000...00 ?

**Comment by zsfelfoldi** [comment]
I think it shouldn't (that's the absolute minimum value), see gitter private channel.

**Pull request merged**